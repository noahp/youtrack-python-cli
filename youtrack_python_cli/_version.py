try:
    from importlib.metadata import PackageNotFoundError
    from importlib.metadata import version as get_version
except ImportError:
    # fallback for python <3.8
    from importlib_metadata import PackageNotFoundError  # type: ignore[assignment]
    from importlib_metadata import version as get_version

# when running tests on the repo, provide a fallback value, since the
# memfault-cli package is not installed at that time
try:
    version = get_version(__package__)
except PackageNotFoundError:
    version = "dev"
