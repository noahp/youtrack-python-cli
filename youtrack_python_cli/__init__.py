"""
Add the version property to the package.
"""

from ._version import VERSION as __version__

__all__ = [
    "__version__",
]
