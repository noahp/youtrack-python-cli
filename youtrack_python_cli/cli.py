"""
CLI interface.
"""

import dataclasses
import http
import json
import logging
import os
import re
import subprocess
from typing import Tuple

import click
from rich import print_json
from rich.console import Console
from rich.table import Table
from you_track_rest_api_client import AuthenticatedClient
from you_track_rest_api_client.api.default import get_issues_id

# from you_track_rest_api_client.models import MyDataModel
from you_track_rest_api_client.types import Response

from ._version import VERSION


class Issue(dict):
    """
    Issue data class, used to print a table of issue data.
    """

    def __init__(self, ticket, issue: dict):
        self.ticket = ticket
        items_to_keep = [
            "idReadable",
            "summary",
            "assignee_name",
            "reporter_name",
            "description",
            "url",
        ]
        filtered_dict = {k: v for k, v in issue.items() if k in items_to_keep}
        filtered_dict = dict(
            sorted(
                filtered_dict.items(),
                key=lambda pair: items_to_keep.index(pair[0]),
            )
        )
        # truncate description to 1024 chars
        if (
            filtered_dict.get("description")
            and len(filtered_dict["description"]) > 1024
        ):
            filtered_dict["description"] = (
                filtered_dict["description"][:1024] + "..."
            )
        super().__init__(filtered_dict)

    def print_table(self):
        """Print a pretty table of the issue data."""
        table = Table(title=f"Issue data for {self.ticket}")

        table.add_column("Key", justify="right", style="cyan", no_wrap=True)
        table.add_column("Value", style="magenta", overflow="fold")
        for key, value in self.items():
            table.add_row(key, value)
            table.add_section()

        # force rich to display fixed width when CI environment variable is set
        # the other approaches (TERM=dumb etc) were not working in the tests.
        if os.environ.get("CI"):
            console = Console(width=80, height=50)
        else:
            console = Console()
        console.print(table)


@dataclasses.dataclass
class CliCtx:
    """
    Context object for the CLI, forwarded from the top-level command group
    """

    url: str
    youtrack: AuthenticatedClient
    verbose: bool


def get_config(key: str) -> str:
    """
    Get a config value from git config, or environment variables.
    """
    config = None

    # try git config
    try:
        config = (
            subprocess.check_output(["git", "config", f"youtrack.{key}"])
            .decode("utf-8")
            .strip()
        )
    except subprocess.CalledProcessError:
        pass

    # try environment variable- overrides git config
    env_var = f"YOUTRACK_{key.upper()}"
    if env_var in os.environ:
        config = os.environ[env_var]

    if not config:
        raise click.UsageError(
            f"missing config value for '{key}', please set via environment"
            f" variable '{env_var}', or via 'git config youtrack.{key}'"
        )

    return config


def load_config(url: str, token: str) -> Tuple[str, str]:
    """
    Load configs.
    """
    return url or get_config("url"), token or get_config("token")


@click.group()
@click.option("--verbose", help="Log verbosely", is_flag=True)
@click.option(
    "--url",
    help=(
        "Explicit youtrack url. Can be configured here, or via environment"
        " variable YOUTRACK_URL, or via 'git config youtrack.url'."
    ),
)
@click.option(
    "--token",
    help=(
        "Explicit youtrack token. Can be configured here, or via environment"
        " variable YOUTRACK_TOKEN, or via 'git config youtrack.token'."
    ),
)
@click.version_option(version=VERSION)
@click.pass_context
def cli(ctx, verbose, url, token):
    """
    Top-level command group, for global args.
    """
    if verbose:
        click.echo(f"version: {VERSION}")

        logging.basicConfig(level=logging.DEBUG)

    url, token = load_config(url, token)

    ctx.obj = CliCtx(
        url, AuthenticatedClient(base_url=url, token=token), verbose
    )


@cli.command()
@click.option(
    "--ticket", help="Ticket id, of the form PROJECT-1234", required=True
)
@click.option(
    "--confirm-prompt",
    help=(
        "Prompt the user to type in the issue, can be used as part of a git"
        " pre-push script for confirming issue"
    ),
    is_flag=True,
)
@click.pass_obj
def get(ctx: CliCtx, ticket, confirm_prompt):
    """
    Fetch a ticket by id. Optionally prompt the user to confirm the ticket id,
    useful for git pre-push hooks.
    """
    if not re.match(r"^[A-Za-z]+-\d+$", ticket):
        raise click.BadParameter("ticket must be of the form PROJECT-1234")

    # issue api request
    ticket_response: Response = get_issues_id.sync_detailed(
        id=ticket, client=ctx.youtrack
    )

    # check for response error
    if ticket_response.status_code != http.HTTPStatus.OK:
        raise click.UsageError(
            "unexpected HTTP status code, check your youtrack token:"
            f" {ticket_response.status_code}"
        )

    # parse json
    response_text = ticket_response.content.decode("utf-8")
    if ctx.verbose:
        print_json(response_text)
    issue_json = json.loads(response_text)

    # inject some additional fields that the Issue() class uses
    # 1. ticket URL
    issue_json[
        "url"
    ] = f"{ctx.url.partition('/api')[0]}/issue/{issue_json['idReadable']}"
    # 2. reporter name
    issue_json["reporter_name"] = issue_json["reporter"]["login"]
    # 3. assignee name, if present
    for field in issue_json["customFields"]:
        if field["name"] == "Assignee" and field["value"]:
            issue_json["assignee_name"] = field["value"]["name"]

    issue = Issue(ticket, issue_json)
    issue.print_table()

    if confirm_prompt:
        while (
            click.prompt("Type the ticket id to confirm").lower()
            != ticket.lower()
        ):
            print("Ticket id did not match, try again!")
