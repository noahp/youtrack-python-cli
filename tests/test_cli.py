import subprocess

import pytest
from click.testing import CliRunner

from youtrack_python_cli.cli import cli


def test_version():
    runner = CliRunner()
    result = runner.invoke(
        cli,
        ["--version"],
    )
    assert result.exit_code == 0
    assert result.output.startswith("cli, version ")


def test_help(snapshot):
    runner = CliRunner()
    result = runner.invoke(
        cli,
        ["--help"],
    )
    assert result.exit_code == 0
    assert result.output == snapshot


BASIC_RESPONSE_JSON = {
    "idReadable": "TEST-1234",
    "reporter": {"login": "testuser"},
    "customFields": [
        {
            "name": "Assignee",
            "value": {
                "name": "testuser name",
            },
        },
        {
            "name": "NotAssignee",
            "value": {
                "name": "testuser not name",
            },
        },
    ],
    "description": "Test Description",
}
BASIC_RESPONSE_JSON_NO_ASSIGNEE = {
    "idReadable": "TEST-1234",
    "reporter": {"login": "testuser"},
    "customFields": [
        {
            "name": "Assignee",
            "value": None,
        },
    ],
    "description": "Test Description",
}
LONG_DESCRIPTION_JSON = BASIC_RESPONSE_JSON.copy()
LONG_DESCRIPTION_JSON["description"] = "a" * 1025
BASIC_CLI_ARGS = [
    "--url",
    "http://youtrack-test/youtrack/api",
    "--token",
    "abc",
]
BASIC_GET_ARGS = BASIC_CLI_ARGS + [
    "get",
    "--ticket",
    "TEST-1234",
]
BASIC_REQUEST_MOCK_URL = r"http://youtrack-test/youtrack/api/issues/TEST-1234?fields=%24type%2Ccreated%2CcustomFields%28%24type%2Cid%2Cname%2Cvalue%28%24type%2Cid%2Cname%29%29%2Cdescription%2Cid%2CidReadable%2Clinks%28%24type%2Cdirection%2Cid%2ClinkType%28%24type%2Cid%2ClocalizedName%2Cname%29%29%2CnumberInProject%2Cproject%28%24type%2Cid%2Cname%2CshortName%29%2Creporter%28%24type%2Cid%2Clogin%2CringId%29%2Cresolved%2Csummary%2Cupdated%2Cupdater%28%24type%2Cid%2Clogin%2CringId%29%2Cvisibility%28%24type%2Cid%2CpermittedGroups%28%24type%2Cid%2Cname%2CringId%29%2CpermittedUsers%28%24type%2Cid%2Clogin%2CringId%29%29"  # noqa: E501


@pytest.mark.parametrize(
    "json_response,cli_args,retval,http_status_code",
    [
        # use explicit http status codes, snapshots vary depending on python
        # version unfortunately
        (BASIC_RESPONSE_JSON, BASIC_GET_ARGS, 0, 200),
        (
            BASIC_RESPONSE_JSON,
            ["--verbose"] + BASIC_GET_ARGS,
            0,
            200,
        ),
        (LONG_DESCRIPTION_JSON, BASIC_GET_ARGS, 0, 200),
        (
            None,
            BASIC_CLI_ARGS + ["get", "--ticket", "blahblah"],
            2,
            200,
        ),
        (BASIC_RESPONSE_JSON, BASIC_GET_ARGS, 2, 401),
        # test that when there's no assignee, we don't crash
        (BASIC_RESPONSE_JSON_NO_ASSIGNEE, BASIC_GET_ARGS, 0, 200),
    ],
)
def test_basic_operation(
    json_response, cli_args, retval, http_status_code, snapshot, respx_mock
):
    respx_mock.get(BASIC_REQUEST_MOCK_URL).respond(
        status_code=http_status_code,
        json=json_response,
    )

    runner = CliRunner()
    result = runner.invoke(
        cli,
        cli_args,
        # enable CI explicity, though it's normally set in tox. this is in case
        # the user is running pytest manually.
        env={"CI": "1"},
    )
    print(result.output)
    assert result.exit_code == retval
    assert result.output == snapshot


def test_git_config(snapshot, respx_mock):
    respx_mock.get(BASIC_REQUEST_MOCK_URL).respond(
        status_code=200,
        json=BASIC_RESPONSE_JSON,
    )

    runner = CliRunner()

    with runner.isolated_filesystem():
        subprocess.check_call(["git", "init"])
        subprocess.check_call(
            [
                "git",
                "config",
                "youtrack.url",
                "http://youtrack-test/youtrack/api",
            ]
        )
        subprocess.check_call(["git", "config", "youtrack.token", "abc"])
        result = runner.invoke(
            cli,
            [
                "get",
                "--ticket",
                "TEST-1234",
            ],
        )
        assert result.exit_code == 0
        assert result.output == snapshot


def test_env_config(snapshot, respx_mock):
    respx_mock.get(BASIC_REQUEST_MOCK_URL).respond(
        status_code=200,
        json=BASIC_RESPONSE_JSON,
    )

    runner = CliRunner()

    result = runner.invoke(
        cli,
        [
            "get",
            "--ticket",
            "TEST-1234",
        ],
        env={
            "YOUTRACK_URL": "http://youtrack-test/youtrack/api",
            "YOUTRACK_TOKEN": "abc",
        },
    )
    assert result.exit_code == 0
    assert result.output == snapshot


def test_no_config_error(snapshot, respx_mock):
    runner = CliRunner()

    result = runner.invoke(
        cli,
        [
            "get",
            "--ticket",
            "TEST-1234",
        ],
    )
    assert result.exit_code == 2
    assert result.output == snapshot


def test_confirm_prompt(snapshot, respx_mock):
    respx_mock.get(BASIC_REQUEST_MOCK_URL).respond(
        status_code=200,
        json=BASIC_RESPONSE_JSON,
    )

    runner = CliRunner()

    result = runner.invoke(
        cli,
        BASIC_GET_ARGS + ["--confirm-prompt"],
        input="n\nTEST-1234",
    )
    assert result.exit_code == 0
    assert result.output == snapshot


def test_no_ci_run(respx_mock):
    respx_mock.get(BASIC_REQUEST_MOCK_URL).respond(
        status_code=200,
        json=BASIC_RESPONSE_JSON,
    )

    runner = CliRunner()

    result = runner.invoke(
        cli,
        BASIC_GET_ARGS,
        env={"CI": ""},
    )
    assert result.exit_code == 0
    # don't care about the output.
