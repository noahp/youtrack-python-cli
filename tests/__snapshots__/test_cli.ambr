# serializer version: 1
# name: test_basic_operation
  '''
                      Issue data for TEST-1234                     
  ┏━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┓
  ┃           Key ┃ Value                                         ┃
  ┡━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┩
  │    idReadable │ TEST-1234                                     │
  ├───────────────┼───────────────────────────────────────────────┤
  │ reporter_name │ testuser                                      │
  ├───────────────┼───────────────────────────────────────────────┤
  │   description │ Test Description                              │
  ├───────────────┼───────────────────────────────────────────────┤
  │           url │ http://youtrack-test/youtrack/issue/TEST-1234 │
  └───────────────┴───────────────────────────────────────────────┘
  
  '''
# ---
# name: test_help
  '''
  Usage: cli [OPTIONS] COMMAND [ARGS]...
  
    Top-level command group, for global args.
  
  Options:
    --verbose     Log verbosely
    --url TEXT    Explicit youtrack url. Can be configured here, or via
                  environment variable YOUTRACK_URL, or via 'git config
                  youtrack.url'.
    --token TEXT  Explicit youtrack token. Can be configured here, or via
                  environment variable YOUTRACK_TOKEN, or via 'git config
                  youtrack.token'.
    --version     Show the version and exit.
    --help        Show this message and exit.
  
  Commands:
    get  Fetch a ticket by id.
  
  '''
# ---
# name: test_version
  '''
  cli, version 0.1.0
  
  '''
# ---
